jmolecules-architecture/jmolecules-cqrs-architecture/src/main/java/org/jmolecules/architecture/cqrs/annotation/CommandPublisher.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jmolecules.architecture.cqrs.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Identifies a command publisher in the context of CQRS, i.e. logic to publish a {@link Command}.
+ *
+ * @author Christian Stettler
+ * @author Henning Schwentner
+ * @author Stephan Pirnbaum
+ * @author Martin Schimak
+ * @author Oliver Drotbohm
+ * @see <a href="http://cqrs.files.wordpress.com/2010/11/cqrs_documents.pdf">CQRS Documents by Greg Young - Commands</a>
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Documented
+public @interface CommandPublisher {
```

Comment:
Do you "publish" a command? Or do you rather "issue" it? Same questions that apply to the `@DomainEventPublisher` regarding general feasibility and whether this should be a type or method annotation. See below.

Reply:
Interesting question. I think you would issue a command but Publisher is far more used und better undersood by developers.

Reply:
Generally we tend to use the words `dispatch` or `send`. In this context, a `CommandDispatcher` might be an option.
To me, Publish sounds like it is sent to a broad audience, which is not the case for Commands.

Reply:
It is unclear to me where this annotation would be placed. Is it placed on the method that is expected to send a command as part of its execution, or is it placed on the method that will actually do the dispatching. In an Axon context, the latter would be a method on the CommandGateway and/or CommandBus, for example.
I would prefer the latter because otherwise, the annotations would need to give away implementation details of the method it annotated. Also, I am not sure if the "publishes" attribute here would be very useful. That may vary at runtime, and I suggest relying on the "type" of the actual class passed as a parameter instead.

Reply:
Let me detail this out a bit:

The publisher annotation is indeed stating information that can be identified from the source code. So imagine an application that implements some kind of a generic message sender interface or always makes use of e.g. a KafkaTemplate. 

Since I'm interested in where specific events are published, I wouldn't care for the generic interface, but for the places where this is invoked and for which events this happens.  External tools such as jQAssistant cannot by default identify all possible ways how event publishing can be implemented, leading to additional effort for the user. Plus, figuring out what kind of events is used to invoke a generic method is another level of complexity. On the contrary, having this annotation, would allow for this and might allow frameworks or custom code to plugin further functionality. 

I agree to the point that this leads to more verbose code. But in the end this is an individual decision to be taken by the architects/developers. If they decide to use it, they'll earn additional functionality from external tools or other frameworks, and if they don't want to use it, they simply won't have that benefit.

Reply:
Okay, that makes sense to me assuming were especially if we're talking about event publishers. I suggest we add a note to the Javadoc of those annotations that they rather serve identification purposes and might only be necessary to be added *if* you'd like to use tools like jQA to discover those places.

I can also imagine our technology integration to take care of the "annotating". The current implementations already use ByteBuddy to map jMolecules annotations to e.g. Spring ones. We can certainly invert this as well to avoid the tools integrating jMolecules having to know about technology specifics. E.g. within the Spring specific one, we could find the aforementioned invocations of publishing methods on a `KafkaTemplate` and add the annotation in the enhancement step so that jQA would find the annotation even thought it wasn't explicitly added by the user.

That said, I wonder if – especially for the `@EventPublisher` – it'd make sense to add a `INTERNAL` / `EXTERNAL` / `UNKNOWN` (default) flag as I guess it makes sense to distinguish between event publishers that explicitly externalize events (like the Kafka example) VS. ones that create them for VM internal consumption. As Axon implements location transparency sticking with a default of `UNKNOWN` would simply leave that undecided. Those flags could then be evaluated when creating documentation or implying rules on them. Did we already agree on dropping the `…Publisher` annotations for this round? If so, I'd postpone that idea to the ticket that'll discuss this further. If not, does it make sense to add that right away or shall we discuss this separately, too?

Reply:
I like the idea of providing the type as this allows to explicitly distinguish between internal and external events and, subsequntially, helps to identify external communication e.g. for documentation purpose. As such, I provided an implementation for disucssion within this PR.